core/resource/AbstractCreateOrReuse.php

```diff
@@ -0,0 +1,141 @@
+<?php
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2016 (original work) Open Assessment Technologies SA
+ * 
+ */
+
+namespace oat\generis\model\resource;
+
+use core_kernel_classes_Class;
+use core_kernel_classes_Resource;
+use oat\generis\model\kernel\persistence\smoothsql\search\ComplexSearchService;
+use oat\generis\model\kernel\persistence\smoothsql\search\TaoResultSet;
+use oat\generis\model\resource\exception\DuplicateResourceException;
+use oat\oatbox\service\ConfigurableService;
+use oat\generis\model\OntologyAwareTrait;
+
+/**
+ * Abstract base of CreateAndReuse service
+ *
+ * @author Christophe GARCIA <christopheg@taotesting.com>
+ */
+abstract class AbstractCreateOrReuse 
+    extends ConfigurableService
+    implements CreateOrReuseInterface
+{
+    use OntologyAwareTrait;
+    
+    /**
+     * Returns the common parent all resources have
+     *
+     * @return \core_kernel_classes_Class
+     */
+    abstract public function getParentClass();
+    
+    /**
+     * List of keys that need to be identical between
+     * two resources to represent equivalence
+     *
+     * @return string[]
+     */
+    abstract public function getUniquePredicates();
+
+    /**
+     * WILL break on non smooth implementations
+     *
+     * @return ComplexSearchService
+     */
+    protected function getSearchService() {
+        return $this->getModel()->getSearchInterface();
+    }
+    
+    /**
+     * 
+     * @param array $values
+     * @return TaoResultSet
+     */
+    protected function searchResource(array $values) {
+        
+        $searchService = $this->getSearchService();
```

Comment:
You should use Model abstraction and not directly call search service (smoothmodel)